### PR TITLE
[SPARK-11994][Mllib] Word2VecModel load and save cause SparkException when model is bigger than spark.kryoserializer.buffer.max

### PR DESCRIPTION
**Pull request by tmnd1991**

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala [hunk 1]
```diff
@@ -600,12 +600,15 @@ object Word2VecModel extends Loader[Word2VecModel] {
       val vectorSize = model.values.head.size
       val numWords = model.size
       val metadata = compact(render
-        (("class" -> classNameV1_0) ~ ("version" -> formatVersionV1_0) ~
-         ("vectorSize" -> vectorSize) ~ ("numWords" -> numWords)))
+      (("class" -> classNameV1_0) ~ ("version" -> formatVersionV1_0) ~
+        ("vectorSize" -> vectorSize) ~ ("numWords" -> numWords)))
       sc.parallelize(Seq(metadata), 1).saveAsTextFile(Loader.metadataPath(path))
 
+      val partitionSize = 33554432l //32MB
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala]: `1 << 25`? Can you explain the rationale in a quick comment here, and what 41 comes from?

**Comment by tmnd1991** [comment]
I explained it on the Jira issue, I'm going to explain it again here:
Since `spark.kryoserializer.buffer.max` defaults to 64MB, I decided to increase the number of partitions the model gets divided into at half that size (32MB).
One word2vec entry consists of an array of float of size vectorSize and a string, since the size of string is variable and considerably lower than the size of the array, I'm not going consider it in my size computation.
The number of partitions the model gets splitted into is given by the formulae: 
`(4 * numWords * vectorSize / 33554432) + 1`
Where 4 is float size, numberWords and vectorSize are respectively the number of words the model contains and the size of each array, and 33554432 is 32MB in bytes.
One more sophisticated solution would be to read the `spark.kryoserializer.buffer.max` value at runtime, but it would be kinda meaningless, because when we're saving the model we're not sure that the property value will be the same when we load it. It can be a different Spark application.

**Comment by srowen** [comment]
@tmnd1991 Yes, I know. I'm suggesting you explain it not in the JIRA or PR, but briefly as comments in the code.

`1 << 25` is also 32MB in that it's 2^25. It seems like less of a magic number than "33554432". (Especially since you made it a long but with an 'l' instead of 'L' which makes it look like the number 335544321. Up to your taste though.

**Comment by tmnd1991** [comment]
I adjusted the code as you suggested

**Comment by srowen** [comment]
This is tagged to the wrong JIRA right? It's SPARK-11994

**Comment by tmnd1991** [comment]
Yes, it was wrong, I fixed it, sorry.

**Comment by tmnd1991** [comment]
I adjusted the code style.

**Comment by srowen** [comment]
(The title is still wrong, it's SPARK-11994)

**Comment by tmnd1991** [comment]
Something went wrong with the commit.
Now should be fine. Never commit before a good coffee!

**Comment by srowen** [comment]
@MechCoder @rxin any thoughts on this one? looks reasonable unless there's some reason it has to be in 1 partition.

**Comment by srowen** [comment]
Checking with @mengxr @jkbradley too just in case

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**